### ISSUE TITLE
Investigate/Document instructions needed for simulation mode

### ISSUE THREAD

**Issue by anakrish** [openenclave/openenclave#1770]
Open Enclave SDK requires xsave, fsgsbase etc for simulation mode.
The goal of this issue to document the specific set of instructions that are needed to run enclaves in simulation mode.

@shruti25ratnam @jxyang Can you investigate creating VMs using KVM and/or VirtualBox and see which instructions need to be enabled in the VMs for simulation mode, and update the documentation.
CC: @aserdean since you were also interested in getting into OE core.

**Comment by CodeMonkeyLeet** [comment]
@shruti25ratnam Can you update this issue?

**Comment by shruti25ratnam** [comment]
I quickly tried creating a VM via Hyper-V and I didn't notice any ctest failures. Is this only specific to VirtualBox/KVM?
![image](https://user-images.githubusercontent.com/7614310/58906385-9f8a6200-86c0-11e9-8894-a53793fa4f21.png)

**Comment by yakman2020** [comment]
Is this still active/accurate? Do we want to continue supporting simulation mode?

**Comment by dthaler** [comment]
@anakrish what is the current status of updating the docs?

For example PR #2600 description says docs were deferred to a separate PR, and I see building_oe_sdk.md still says "On Windows, Open Enclave SDK does not support simulation mode."

**Comment by anakrish** [comment]
@radhikaj Do we want to claim simulation mode support on Windows? Though it has parity with Linux simulation mode and debuggers work too, I am a bit hesitant because currently you can build/run OE in simulation mode on Windows only on SGX machines; whereas you can build, run OE in simulation mode on non-sgx machines.

What do you think?

**Comment by dthaler** [comment]
Then I think we should claim simulation mode support on SGX machines, but not on non-SGX machines.   (I would like it to work on non-SGX machines too, but this issue is docing what is supported.)

**Comment by achamayou** [comment]
The main (only?) thing that prevents simulation mode from working effectively on non-SGX kit is the fact that oesign creates an enclave.

That makes signing easier but there doesn’t seem to be any fundamental reason why it couldn’t be done outside an enclave (presumably it’s a matter of knowing exactly what the memory layout is and how to produce the digest). If that restriction was lifted, then simulation mode could become usable on non-SGX machines too.

Another thing that’s awkward in simulation mode but not a complete blocker is the fact that the attestation code doesn’t work, rather than produce/verify stub quotes, so applications must know they’re run in simulation mode and turn off a bunch of code paths and features that depend on attestation.

**Comment by dthaler** [comment]
https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Simulation.md does say that simulation mode will work (without attestation, etc.) on non-SGX machines but only on Linux, although it's not clear why Linux vs Windows would make a difference.

**Comment by anakrish** [comment]
> https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Simulation.md does say that simulation mode will work (without attestation, etc.)

That was my understanding too. That simulation mode will work on non-SGX machines on Linux.

I don't think that is accurate anymore. This is what ldd reveals when the SDK has been built with HAS_QUOTE_PROVIDER=off (which means no attestation support):

```bash
build/tests/echo/host$ ldd echo_host 
	linux-vdso.so.1 (0x00007fff507ca000)
	libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007fa3e0664000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fa3e065e000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa3e063b000)

	libsgx_enclave_common.so.1 => /usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1 (0x00007fa3e0633000)

	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa3e0442000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa3e095c000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fa3e0254000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fa3e0103000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fa3e00e9000)
```

We seem to have a hard dependency on SGX library (libsgx_enclave_common.so). So right now, I think simulation mode will work only on SGX machines on both Linux and Windows.

**Comment by anakrish** [comment]
Ok. Confirmed that we are consistent on both Windows and Linux in terms of hard dependency on sgx libraries.

https://github.com/openenclave/openenclave/blob/6bed4b5e0927746ca2e9f0c0c8b154e5dd66a204/host/CMakeLists.txt#L428-L440

**Comment by anakrish** [comment]
Going by git history, unless I'm mistaken, this hard dependency seems to have been introduced during Oct, 2019: https://github.com/openenclave/openenclave/commit/7e1072825796e112efd5bfe38db3f20d8390cd82

**Comment by anakrish** [comment]
> Then I think we should claim simulation mode support on SGX machines, but not on non-SGX machines.

This is accurate and what we need to change the doc to.

**Comment by jhand2** [comment]
You can install libsgx_enclave_common on non-SGX machines. I use it all the time on an non-SGX linux machine. I can't speak to Windows but I know it works as expected on Linux.

**Comment by anakrish** [comment]
@jhand2 How thoroughly has your approach been tested? If it is robust, can you go ahead and update the docs to include your solution?

**Comment by jhand2** [comment]
I've been using it since November. I can try to get to it after my other PRs for 0.9 get merged.

**Comment by achamayou** [comment]
@jhand2 are you able to run oesign on non-SGX kit under Linux?

**Comment by jhand2** [comment]
I am. Also it appears when building with -DHAS_QUOTE_PROVIDER=OFF oesign has no libsgx dependencies:

```
jordan@jordan-xps $ ldd ./oesign
        linux-vdso.so.1 (0x00007ffd76a72000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007faa810b4000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007faa81092000)
        libc.so.6 => /lib64/libc.so.6 (0x00007faa80ec9000)
        libz.so.1 => /lib64/libz.so.1 (0x00007faa80eaf000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007faa80ea8000)
        /lib64/ld-linux-x86-64.so.2 (0x00007faa813c4000)
```

**Comment by jhand2** [comment]
Note that I run a fedora machine so none of the Intel PSW is installed via packages. I just built it from source. I could probably scrounge up a non-SGX ubuntu machine and see if the SGX packages still install as expected or if you would have to build the PSW from source there as well.

**Comment by achamayou** [comment]
I'm surprised because oesign calls oe_sgx_build_enclave, and that issues an ECREATE it seems: https://github.com/openenclave/openenclave/blob/b531f129c12c5bccd61da9bb797bd897397f7971/host/sgx/create.c#L655

Perhaps it no longer does, I'll try it on a non-sgx machine.

**Comment by jhand2** [comment]
@achamayou oesign links against oehostmr (rather than oehost). oehostmr ifdef's out all the calls to libsgx functions and just does measurement manually.

For example, enclave_create call is note compiled into oesign:

https://github.com/openenclave/openenclave/blob/991d6d39e18519133b441be99ea28a7dd245e6d9/host/sgx/sgxload.c#L452-L479

**Comment by radhikaj** [comment]
> @achamayou oesign links against oehostmr (rather than oehost). oehostmr ifdef's out all the calls to libsgx functions and just does measurement manually.
> 
> For example, enclave_create call is note compiled into oesign:
> 
> https://github.com/openenclave/openenclave/blob/991d6d39e18519133b441be99ea28a7dd245e6d9/host/sgx/sgxload.c#L452-L479

Specifically, after https://github.com/openenclave/openenclave/pull/2644 was merged, oesign has no dependencies on intel sgx libs.

**Comment by achamayou** [comment]
@radhikaj @jhand2 thank you for this clarification, I was not aware this had been fixed, it’s excellent news :)

**Comment by radhikaj** [comment]
Fixed in #2934 2874
